cassiopeia/spatial/spatial_imputation.py

```diff
@@ -0,0 +1,125 @@
+"""
+
+from typing import Optional
+
+import anndata
+import networkx as nx
+import numpy as np
+import pandas as pd
+import tqdm
+
+from cassiopeia.spatial import spatial_utilities
+
+
+def impute_spatial_data(
+    character_matrix: pd.DataFrame,
+    adata: Optional[anndata.AnnData] = None,
+    spatial_graph: Optional[nx.Graph] = None,
+    neighborhood_size: Optional[int] = None,
+    neighborhood_radius: float = 30.0,
+    imputation_hops: int = 2,
+    imputation_concordance: float = 0.8,
+    num_imputation_iterations: int = 1,
+    max_neighbor_distance: float = np.inf,
+    coordinates: Optional[pd.DataFrame] = None,
+) -> pd.DataFrame:
+    """Imputes data based on spatial location.
+
+    This procedure iterates over spots in a given anndata and imputes missing
+    data based on spatial neigborhoods. This procedure is iterative, and can act
+    as an approximation of a message-passing process to smooth over missing
+    data.
+
+    Args:
+        character_matrix: A character matrix of spots, constructed using a
+            function like `convert_allele_table_to_character_matrix`.
+        adata: Anndata of spatially-resolved data. Only the spatial coordinates
+            need to be stored, and this is used to construct a graph.
+        spatial_graph: Optionally, the user can provide a spatial connectivity
+            graph instead of passing in an adata.
+        neighborhood_size: If a connectivitity graph is being constructed,
+            this is the number of nearest neighbors to connect to a node. If
+            both neighborhood_size and neighborhood_radius are passed in,
+            neighborhood_size is preferred.
+        neighborhood_radius: Intead of passing in `neighborhood_size`, this
+            is the radius of the connectivity graph.
+        imputation_hops: Number of adjacent node's adjacencies to query. For
+            example, if this is 2, this means that imputation is done not just
+            on nearest neighbors of a given node, but also each nearest
+            neighbor's nearest neighbors.
+        imputation_concordance: Fraction of votes that must agree in order
+            to accept an imputation.
+        num_imputation_iterations: Number of iterations for imputation
+            procedure.
+        max_neighbor_distance: Maximum distance to neighbor to be used for
+            imputation.
+        coordinates: If an AnnData is not specified, and you wish to set an
+            upper limit on the distance for spatial imputation, these
+            coordinates can be passed to the imputation procedure.
+
+    Returns:
+        An imputed character matrix.
+    """
+    if (not spatial_graph) and (not adata):
+        raise Exception(
+            "One of the following must be specified: "
+            "`spatial_graph` or `adata`."
+        )
+
+    if not spatial_graph:
+
+        # create spatial graph if needed
+        spatial_graph = spatial_utilities.get_spatial_graph_from_anndata(
```

Comment:
I suggest requiring that the spatial_graph is precomputed to eliminate the squidpy dependency

Reply:
I propose we keep it this way for now until we have the treedata class as a dependency in which case a spatial graph constructor will presumably be implemented.